### PR TITLE
Fix MTP CP-boundary roll to communicate on the context-parallel process group

### PR DESCRIPTION
**Pull request by liding-nv**
# What does this PR do?

Fixes MTP's CP-boundary roll (`roll_tensor` / `_roll_tensor_packed_seq`) to communicate on the context-parallel process group instead of the default (world) process group.

## Details

Both roll variants compute their neighbor ranks from `cp_group` but issue `isend`/`irecv` without `group=`, so every CP boundary exchange actually runs on the default process group. When the world equals the CP group (typical unit tests) this is invisible. Whenever the CP group is a strict subgroup of the world (TP > 1 and/or multiple DP replicas), all CP rings in the job share one communicator:

- eager-init unbatched P2P is serialized against every other op on that group — PyTorch warns: *"An unbatched P2P op (send/recv) was called on this ProcessGroup … serialized with all other ops on this ProcessGroup"*;
- independent rings' P2P schedules couple through the shared communicator, a hang hazard that grows with world size — especially with packed sequences, where `_roll_tensor_packed_seq` posts the exchange per packed sequence per microbatch.

The fix addresses the ops to `cp_group` and batches each rank's sends/receives with `batch_isend_irecv`.

## Testing

- New unit tests (`TestRollTensorWithCPSubgroup`) run both roll variants on a CP group that is a strict subgroup of the world (TP2 × CP2), asserting numeric equivalence with the single-rank semantics and that the exchange is addressed to the CP group (the latter fails on the pre-fix code).
- Full `test_multi_token_prediction.py` run before/after the fix on 4 GPUs produces identical results (failing cases on 4 GPUs are pre-existing world-size/fp8 environment mismatches, unchanged by this PR).
- Loss parity verified in a 5-node MTP+CP training A/B (fixed vs unfixed): loss trajectories match to ~1e-6 relative.
- `tools/autoformat.sh` run; black/isort/pylint/ruff clean (mypy findings are pre-existing, outside the changed lines).

### CODE REVIEW

**Review by yashaswikarnati — APPROVED** [review]

**Review by jaredcasper — APPROVED** [review]

File: tests/unit_tests/transformer/test_multi_token_prediction.py [hunk 1]
```diff
@@ -3931,6 +3931,90 @@ def test_attention_mask_validation_mamba(self):
                 raise
 
 
+class TestRollTensorWithCPSubgroup:
+    """Roll-tensor CP paths on a CP group that is a strict subgroup of WORLD.
```
Comment (liding-nv) [inline tests/unit_tests/transformer/test_multi_token_prediction.py:3935]: The existing packed-sequence test uses TP1 × CP2 and checks numerical correctness without checking communicator selection. When run with two ranks, CP and WORLD contain the same ranks, hiding the communicator mismatch from numerical checks.

This test uses TP2 × CP2 to guarantee CP is a strict subgroup of WORLD, and explicitly checks that boundary exchanges use cp_group. The explicit group assertion detects the regression

File: tests/unit_tests/transformer/test_multi_token_prediction.py [hunk 2]
```diff
@@ -3931,6 +3931,90 @@ def test_attention_mask_validation_mamba(self):
...
+        def spy_batch_isend_irecv(ops):
+            seen_groups.extend(op.group for op in ops)
+            return real_batch_isend_irecv(ops)
+
+        monkeypatch.setattr(torch.distributed, "batch_isend_irecv", spy_batch_isend_irecv)
+
+        full = torch.arange(2 * 16, dtype=torch.float32, device="cuda").reshape(2, 16)
+        # Oracle: the documented single-rank semantics on the full sequence.
+        expected_full, _ = roll_tensor(full.clone(), shifts=-1, dims=-1)
+        expected_local = self._cp_shard(expected_full, cp_rank, cp_size)
```
Comment (liding-nv) [inline tests/unit_tests/transformer/test_multi_token_prediction.py:3974]: We build the reference by rolling the full sequence without distributed communication, then extracting this rank’s CP shard. 
The actual path shards the input first, then rolls it with cross-rank communication. Both paths should produce identical local values.

File: tests/unit_tests/transformer/test_multi_token_prediction.py [hunk 3]
```diff
@@ -3931,6 +3931,90 @@ def test_attention_mask_validation_mamba(self):
...
+
+        local = self._cp_shard(full, cp_rank, cp_size).contiguous()
+        rolled, rolled_sum = roll_tensor(local, shifts=-1, dims=-1, cp_group=cp_group)
+
+        assert torch.equal(rolled, expected_local)
+        assert rolled_sum == expected_local.sum()
```
Comment (liding-nv) [inline tests/unit_tests/transformer/test_multi_token_prediction.py:3980]: These check the local output values—including exchanged boundary tokens and zero padding—and the returned local sum. They verify numerical behavior is preserved. They can also pass before the fix because the old implementation communicates with the correct peers through WORLD.

**Review by zhongbozhu — APPROVED** [review]

**Review by cuichenx — APPROVED** [review]
